internal_worker_base.go

```diff
@@ -121,18 +135,137 @@ func (bw *baseWorker) Start() {
 		return
 	}
 
-	// Launch the routines to do work
-	for i := 0; i < bw.options.routineCount; i++ {
+	for i := 0; i < bw.options.pollerCount; i++ {
 		bw.shutdownWG.Add(1)
-		go bw.execute(i)
+		go bw.runPoller()
 	}
+	bw.shutdownWG.Add(1)
+	go bw.runTaskDispatcher()
 
 	bw.isWorkerStarted = true
 	traceLog(func() {
-		bw.logger.Info("Started Worker", zap.Int("RoutineCount", bw.options.routineCount))
+		bw.logger.Info("Started Worker",
+			zap.Int("PollerCount", bw.options.pollerCount),
+			zap.Int("MaxConcurrentTask", bw.options.maxConcurrentTask),
+			zap.Int("MaxTaskRps", bw.options.maxTaskRps),
+		)
 	})
 }
 
+func (bw *baseWorker) isShutdown() bool {
+	select {
+	case <-bw.shutdownCh:
+		return true
+	default:
+		return false
+	}
+}
+
+func (bw *baseWorker) runPoller() {
+	defer bw.shutdownWG.Done()
+
+	bw.pollerRequestCh <- struct{}{} // initial poll request to kick off the poller loop
+
+	for {
+		select {
+		case <-bw.shutdownCh:
+			return
+		case <-bw.pollerRequestCh:
+			ch := make(chan struct{})
+			go func(ch chan struct{}) {
+				bw.pollTask()
+				close(ch)
+			}(ch)
+
+			// block until previous poll completed or return immediately when shutdown
+			select {
+			case <-bw.shutdownCh:
+				return
+			case <-ch:
+			}
+		}
+	}
+}
+
+func (bw *baseWorker) runTaskDispatcher() {
+	defer bw.shutdownWG.Done()
+
+	for i := 0; i < bw.options.maxConcurrentTask; i++ {
+		bw.taskSemaphore <- struct{}{}
+	}
+
+	for {
+		// wait for new task or shutdown
+		select {
+		case <-bw.shutdownCh:
+			return
+		case task := <-bw.taskQueueCh:
+			// got new task, wait for taskSemaphore or shutdown
+			select {
+			case <-bw.shutdownCh:
+				return
+			case <-bw.taskSemaphore:
+				// block until taskRateLimiter satisfied
+				for !bw.taskRateLimiter.Consume(1, time.Millisecond) {
+					if bw.isShutdown() {
+						return
+					}
+				}
+				go bw.processTask(task)
+			}
+		}
+	}
+}
+
+func (bw *baseWorker) pollTask() {
+	var err error
+	var task interface{}
+	bw.retrier.Throttle()
+	if bw.pollRateLimiter.Consume(1, time.Millisecond) {
+		task, err = bw.options.taskWorker.PollTask()
+		if err != nil {
+			bw.retrier.Failed()
+			if enableVerboseLogging {
+				bw.logger.Debug("Failed to poll for task.", zap.Error(err))
+			}
+		} else {
+			bw.retrier.Succeeded()
+		}
+	} else {
+		// if rate limiter fails, we need to back off.
+		bw.retrier.Failed()
+	}
+
+	if task != nil {
+		bw.taskQueueCh <- task
+	} else {
+		bw.pollerRequestCh <- struct{}{} // poll failed, trigger a new pool
+	}
+}
+
+func (bw *baseWorker) processTask(task interface{}) {
+	pollerRequested := false
+	if bw.concurrentTaskCount.Inc() < int32(bw.options.maxConcurrentTask) {
```

Comment:
5 is not true, after worker 1 completes, it will put a request (see end of this method), we don't wait for worker 2, and that is desired behavior because before worker 1 is completed, we don't want poll. The problem of use cap of worker count and pre fill pollRequestCh has the problem that i described earlier that it will result refetch in case of large maxConcurrentTask.